### DISCUSSION TITLE
Failed to locate command: [Errno 2] No such file or directory: 'git'

### DISCUSSION THREAD

**Discussion by bcdady** [ansible/ansible-lint#2859] [Q&A]
##### Summary

The issue of 'git' not being automatically excluded has apparently regressed/reocurred, and so -- when the `git` command is not found, an error message is displayed when running `ansible-lint --version`.

I've just installed ansible-lint v6.10 (in a [python:3-slim container](https://hub.docker.com/layers/library/python/3-slim/images/sha256-7ae2d10e4bdc6f69ba2daf031647568fec08f3191621d7a5c8760abb236d16ab?context=explore)) and see this error when running `ansible-lint --version`.

A similar / same symptom was previously addressed in #2641 / #2633 (reported for ansible-lint ver 6.8.4). 

##### Issue Type

- Bug Report

##### OS / ENVIRONMENT

```console
$ ansible-lint --version

Failed to locate command: [Errno 2] No such file or directory: 'git'
ansible-lint 6.10.0 using ansible 2.13.7
```

Since the prior issue (#2641 in October) related to a "[Fix pip version check](https://github.com/ansible/ansible-lint/pull/2641/commits/38042afe843bb342ef263ade6a5d59a4765f50f8)", I'm including confirmation that a current version of pip is installed and running in this same container.

```console
$ pip --version
pip 22.3.1 from /usr/local/lib/python3.11/site-packages/pip (python 3.11)
```

- ansible installation method: pip
- ansible-lint installation method: pip (pipenv)

It seems the same issue is present in ansible-lint v6.9.1 and v6.8.7

```console
$ pip install 'ansible-lint==6.9.*'
Defaulting to user installation because normal site-packages is not writeable
Collecting ansible-lint==6.9.*
...
Installing collected packages: ansible-lint
Successfully installed ansible-lint-6.9.1

$ ansible-lint --version
Failed to locate command: [Errno 2] No such file or directory: 'git'
ansible-lint 6.9.1 using ansible 2.13.7

$ pip uninstall ansible-lint
Found existing installation: ansible-lint 6.9.1
Uninstalling ansible-lint-6.9.1:
...
Successfully uninstalled ansible-lint-6.9.1

$ pip install 'ansible-lint==6.8.*'
Defaulting to user installation because normal site-packages is not writeable
Collecting ansible-lint==6.8.*
...
Installing collected packages: ansible-lint
Successfully installed ansible-lint-6.8.7

$ ansible-lint --version
Failed to locate command: [Errno 2] No such file or directory: 'git'
ansible-lint 6.8.7 using ansible 2.13.7
```

##### STEPS TO REPRODUCE

```console
# Translated from the 'RUN' commands in my Dockerfile, as a non-root user (named ansible):

python -m pip install 'pip>=22'
pip install 'ansible>=6.0'
pip install 'ansible-lint>=6.0'

```

##### Desired Behavior

Unexpected / unnecesarry error messages are not shown when `ansible-lint --version` is run.

##### Actual Behavior

Unexpected / unnecesarry error messages are shown when `ansible-lint --version` is run.

**Comment by ssbarnea** [comment]
Did you check if this message is displayed on stderr or stdout? If is is on stderr, that is not a bug, if is on stdout, it is a bug.

**Reply by bcdady** [reply]
Good point. I confirmed this is `stderr` output.
I contend that when non-functional parameters (such as `--help` and/or `--version` are passed, the whole code path concerned with checking if `git` is available should be skipped as out-of-context. I'm presuming that any dependency on `git` is specific to other functionality of ansible-lint, and is Not necessary in `--help` and/or `--version` mode.

I've not yet determined the common cause, but I'm seeing symptoms when this `stdout` is displayed that apparently halt/interrupt the `--help` and `--version` output.

**Reply by ssbarnea** [reply]
We included some changes on 6.10.2 that made git mandatory, just for avoiding extra confusions and divergent behavior when git was missing. Still, you will not be able to get the version if git is missing, there were too many internal changed needed in order to allow --version work without git being present.

**Reply by bcdady** [reply]
I see.
The symptoms I initially reported were specific to ver. 6.9, and I didn't notice any obvious change in setup experience for 6.10.
Should there be some warning or other message if/when `ansible-lint` is being installed on an OS that does Not have `git`?

Are any of the discussions or other contexts for these decisions and changes available publicly; is so, could you share any relevant links?

**Comment by bcdady** [comment]
It seems to me we should consider failing, or at least provide a stern warning, when installing `ansible-lint` on a system without `git` installed since it is now a _hard_ dependency.

With versions 6.9 and prior I can run `ansible-lint --help` and `ansible-lint --version` and, in addition to the `git` error message, I still get the expected `help` or `version` output.

At least in this case, I could redirect `stderr` to ignore this warning.

6.8 (tested via `pip install 'ansible-lint==6.8.*'`:
```console
# ansible-lint --version
Failed to locate command: [Errno 2] No such file or directory: 'git'
ansible-lint 6.8.7 using ansible 2.14.1
```
6.9 (tested via `pip install 'ansible-lint==6.9.*'`:
```console
# ansible-lint --version
Failed to locate command: [Errno 2] No such file or directory: 'git'
ansible-lint 6.9.1 using ansible 2.14.1
A new release of ansible-lint is available: 6.9.1 → 6.10.2
```

But now with 6.10, I only get the error message and no `help` nor `version` output.

6.10 (tested via `pip install 'ansible-lint>=6.10'`:
```console
# ansible-lint --help
Failed to find runtime dependency 'git' in PATH
/#

# ansible-lint --version
Failed to find runtime dependency 'git' in PATH
/#
```

**Reply by bcdady** [reply]
In fact, I reproduced the same (expected behavior based on earlier minor versions) behavior in 6.10.0 - 6.10.1, which suggests 6.10.2 should be a major version bump, due to the breaking nature of the `git` dependency. Agreed?
